include/xtensor-zarr/xzarr_chunked_array.hpp

```diff
@@ -0,0 +1,128 @@
+*                                                                          *
+* Distributed under the terms of the BSD 3-Clause License.                 *
+*                                                                          *
+* The full license is in the file LICENSE, distributed with this software. *
+****************************************************************************/
+
+#ifndef XTENSOR_ZARR_CHUNKED_ARRAY_HPP
+#define XTENSOR_ZARR_CHUNKED_ARRAY_HPP
+
+#define HALF_ENABLE_F16C_INTRINSICS 0
+#define HALF_ENABLE_CPP11_LONG_LONG 1
+#define HALF_ENABLE_CPP11_STATIC_ASSERT 1
+#define HALF_ENABLE_CPP11_CONSTEXPR 1
+#define HALF_ENABLE_CPP11_NOEXCEPT 1
+#define HALF_ENABLE_CPP11_USER_LITERALS 1
+#define HALF_ENABLE_CPP11_THREAD_LOCAL 1
+#define HALF_ENABLE_CPP11_TYPE_TRAITS 1
+#define HALF_ENABLE_CPP11_CSTDINT 1
+#define HALF_ENABLE_CPP11_CMATH 1
+#define HALF_ENABLE_CPP11_CFENV 1
+#define HALF_ENABLE_CPP11_HASH 1
+#include "half_float/half.hpp"
+#include "xtensor/xchunked_array.hpp"
+#include "xtensor/xchunk_store_manager.hpp"
+#include "xtensor/xfile_array.hpp"
+#include "xtensor/zarray.hpp"
+#include "xtensor-io/xio_binary.hpp"
+#include "xtensor-io/xio_gzip.hpp"
+#include "xzarr_common.hpp"
+
+namespace xt
+{
+    template <class T, class io_handler, class format_config>
+    zarray get_zarray(std::vector<std::size_t>& shape, std::vector<std::size_t>& chunk_shape, const std::string& path, char separator, const nlohmann::json& attrs, char endianness, format_config&& config, const nlohmann::json& config_json)
+    {
+        config.read_from(config_json);
+        config.big_endian = (endianness == '>');
+        xchunked_array<xchunk_store_manager<xfile_array<T, io_handler>, xzarr_index_path>, xzarr_attrs> a(shape, chunk_shape);
+        a.chunks().set_directory(path.c_str());
+        a.chunks().get_index_path().set_separator(separator);
+        a.chunks().configure_format(config);
+        a.set_attrs(attrs);
+        return zarray(a);
+    }
+
+    template <class store_type, class T>
+    zarray build_chunked_array_impl(const std::string& compressor, std::vector<std::size_t>& shape, std::vector<std::size_t>& chunk_shape, const std::string& path, char separator, const nlohmann::json& attrs, char endianness, nlohmann::json& config)
+    {
+        if (compressor == "binary")
+        {
+            using format_config = xio_binary_config;
+            using io_handler = typename store_type::template io_handler<format_config>;
+            return get_zarray<T, io_handler>(shape, chunk_shape, path, separator, attrs, endianness, format_config(), config);
+        }
+        else if (compressor == "gzip")
+        {
+            using format_config = xio_gzip_config;
+            using io_handler = typename store_type::template io_handler<format_config>;
+            return get_zarray<T, io_handler>(shape, chunk_shape, path, separator, attrs, endianness, format_config(), config);
+        }
```

Comment:
@JohanMabille The compressors are currently hard-coded, but we need to find a mechanism for the user to register its own ones.
A map won't work because each format config is a distinct object, they don't share a common type. Actually, xtensor-io's `dump_file` and `load_file` function overloading works based on this fact.
Any idea how we could do?

Reply:
Sorry I just saw your message and you now have a solution. I like it, it's neat!

Reply:
The problem is that it doesn't scale. When a user registers a new compressor, he has to do it for each data type with:
```cpp
xcompressor_factory<store_type, data_type>::add_compressor(c);
```